8312065: Socket.connect does not timeout as expected when profiling

yanglong1010 opened the pull request:
Hi all,

This pull request contains a fix for [JDK-8312065](https://bugs.openjdk.org/browse/JDK-8312065).

The old SocketImpl is still present in JDK 17, can be enabled with -Djdk.net.usePlainSocketImpl.

I have verified that this problem exists on Linux and macOS, and I feel that it also exists on AIX, 
so I repaired these 3 platforms. Windows implementations are different, so don't have this problem.

I added a platform dependent test.

The test needs to get the OS thread ID of the Java thread. It seems that there is no good Java API to get it directly. The minimum requirement of Linux [/proc/thread-self](https://man7.org/linux/man-pages/man5/proc.5.html#:~:text=/proc/thread%2Dself%20(since%20Linux%203.17)) is Linux 3.17.
The JFR event records the OS thread ID, although this method is cumbersome, it is reliable.

The test also needs to use the kill command to send a signal to the thread. macOS's kill seems to only support process, and I don't have an AIX environment, so this test only supports Linux.

Thanks!

apangin commented on test/jdk/java/net/Socket/B8312065.java:
```diff
@@ -0,0 +1,249 @@
+     * @throws ParseException
+     */
+    private static long getOSThreadId() throws IOException, ParseException {
+        startJFR();
```
You may reuse `libNativeThread` for finding threadId and sending signals. This should make the test a lot simpler.
See `SocketAcceptInterruptTest` and `SocketReadInterruptTest` for example.
JFR is an optional feature, it is not available in Minimal VM.

yanglong1010 replied:
Thanks for your advice. I have updated the test code.

apangin commented on test/jdk/java/net/Socket/B8312065.java:
```diff
@@ -0,0 +1,249 @@
+    static void passed() {
+        success = true;
+    }
+}
```
EOL is missing on the last line

yanglong1010 commented:
Gentle ping @apangin

apangin commented:
@yanglong1010 I've added a minor comment, otherwise the fix looks good to me. But you'll need one more review anyway, as I don't have an official reviewer status.
Please also configure your repository to [enable pre-submit tests](https://github.com/openjdk/jdk17u-dev/pull/1639/checks?check_run_id=16167188975).

yanglong1010 commented on test/jdk/java/net/Socket/B8312065.java:
```diff
@@ -0,0 +1,82 @@
+        t.setDaemon(true);
+        t.start();
+
+        latch.await();
```
It's really useless and has been deleted.

yanglong1010 commented:
> @yanglong1010 I've added a minor comment, otherwise the fix looks good to me. But you'll need one more review anyway, as I don't have an official reviewer status. Please also configure your repository to [enable pre-submit tests](https://github.com/openjdk/jdk17u-dev/pull/1639/checks?check_run_id=16167188975).

Thanks a lot. I have done all your suggestions.

vyommani commented on test/jdk/java/net/Socket/B8312065.java:
```diff
@@ -0,0 +1,76 @@
+            // There is no good way to mock SocketTimeoutException, just assume 192.168.255.255 is not in use
+            socket.connect(new InetSocketAddress("192.168.255.255", 8080), 2000);
+        } catch (SocketTimeoutException e) {
+            System.out.println("Test passed");
```
i will suggest you to provide a check that test 'socket.connect' throws 'SocketTimeoutException'  only after the time expires.

store time just before socket connect and in catch block check if time spent is >= timeout specified.

yanglong1010 replied:
Thanks, updated.

vyommani commented on test/jdk/java/net/Socket/B8312065.java:
```diff
@@ -0,0 +1,76 @@
+            } catch (Exception e) {
+                e.printStackTrace();
+            } finally {
+                System.out.println("Test FAILED");
```
don't you think "Test FAILED" will print always ?

yanglong1010 replied:
Hi vyommani. There are two situations in which this sentence is executed, one is that it loops for 10 seconds (this will not happen on the JDK I repaired), and the other is that the sleep is interrupted (it will not happen).
I ran the test on the JDK containing my fix and watched stdout, the message was indeed not printed and the test passed. On the JDK before the repair, it will print, and the test will fail.
But in order to make this code look clearer, I removed 'finnaly'.

phohensee reviewed (changes requested):
Code duplication is the source of many bugs, so nstead of duplicating the macro code inline, I suggest refactoring a bit and add a RETRY macro argument, vis,
```
#define BLOCKING_IO_RETURN_INT(FD, FUNC, RETRY) {     \
    int ret;                                          \
    threadEntry_t self;                               \
    fdEntry_t *fdEntry = getFdEntry(FD);              \
    if (fdEntry == NULL) {                            \
        errno = EBADF;                                \
        return -1;                                    \
    }                                                 \
    do {                                              \
        startOp(fdEntry, &self);                      \
        ret = FUNC;                                   \
        endOp(fdEntry, &self);                        \
    } while ((RETRY) && ret == -1 && errno == EINTR); \
    return ret;                                       \
}
```
Then, add "true" as the last argument to all invocations of the macro, except the one in NET_Poll, which becomes
```
int NET_Poll(struct pollfd *ufds, unsigned int nfds, int timeout) {
    BLOCKING_IO_RETURN_INT( ufds[0].fd, poll(ufds, nfds, timeout), false );
}
```

phohensee commented:
The GHA failures seem to be infrastructure issues. Your next merge will rerun them, hopefully successfully.

yanglong1010 commented on src/java.base/aix/native/libnet/aix_close.c:
```diff
@@ -489,7 +489,17 @@ int NET_Connect(int s, struct sockaddr *addr, int addrlen) {
+    startOp(fdEntry, &self);
+    ret = poll(ufds, nfds, timeout);
+    endOp(fdEntry, &self);
+    return ret;
```
hi, vyommani, thanks for the suggestion. Do you mind if I modify it like Paul did, I feel like this would make the macro cleaner.

vyommani replied:
yes, that's more cleaner approach.

apangin commented on src/java.base/linux/native/libnet/linux_close.c:
```diff
@@ -40,6 +40,13 @@
+#define true 1
+#define false 0
+#endif
+
```
Since the file already includes `jvm.h`, you can use `JNI_TRUE`/`JNI_FALSE` to avoid introducing more ifdefs

yanglong1010 replied:
Andrei, thanks. Pushed.

yanglong1010 commented:
> Code duplication is the source of many bugs, so instead of duplicating the macro code inline, add a RETRY macro argument, vis,
> 
> ```
> #define false 0
> #define true 1
> 
> #define BLOCKING_IO_RETURN_INT(FD, FUNC, RETRY) {     \
>     int ret;                                          \
>     threadEntry_t self;                               \
>     fdEntry_t *fdEntry = getFdEntry(FD);              \
>     if (fdEntry == NULL) {                            \
>         errno = EBADF;                                \
>         return -1;                                    \
>     }                                                 \
>     do {                                              \
>         startOp(fdEntry, &self);                      \
>         ret = FUNC;                                   \
>         endOp(fdEntry, &self);                        \
>     } while ((RETRY) && ret == -1 && errno == EINTR); \
>     return ret;                                       \
> }
> ```
> 
> Then, add "true" as the last argument to all invocations of the macro, except the one in NET_Poll, which becomes
> 
> ```
> int NET_Poll(struct pollfd *ufds, unsigned int nfds, int timeout) {
>     BLOCKING_IO_RETURN_INT( ufds[0].fd, poll(ufds, nfds, timeout), false );
> }
> ```

Paul, Thank you very much for your detailed advice.

vyommani commented on test/jdk/java/net/Socket/B8312065.java:
```diff
@@ -0,0 +1,87 @@
+        t.setDaemon(true);
+        t.start();
+
+        long startMillis = System.currentTimeMillis();
```
please use monotonically increasing time(System.nanoTime()) in place of 'currentTimeMillis'

yanglong1010 replied:
Thanks, vyommani.  I have pushed. System wall-clock may be changed at any time, It is indeed more reasonable to use `nanoTime`.

vyommani reviewed (commented):
There are couples of formatting error(green bar), i think it is because of tab. Please use space in place of tab.

yanglong1010 commented:
> There are couples of formatting error(green bar), i think it is because of tab. Please use space in place of tab.

vyommani, I carefully checked the files I have modified, and I did not find `tab`s, all the blanks are `space`s.
I think it may be because the line where the `while` keyword is located has become longer. For alignment, I added spaces in adjacent lines.

vyommani reviewed (approved):
Looks OK to me.

GoeLin commented:
Could you please trigger the GHA tests again and the supply a Fix request comment.
I remove the tag in the meantime.

yanglong1010 commented:
> Could you please trigger the GHA tests again and the supply a Fix request comment. I remove the tag in the meantime.

Of course, my pleasure.

yanglong1010 commented:
/approval

yanglong1010 commented:
/approval 8312065 request

apangin commented:
@yanglong1010 I guess you need to merge the latest changes from master to fix GitHub Actions.

GoeLin commented:
You are still missing the fix request comment that states reason, risk and testing in the JBS bug.
You can issue another approval request to do that, or edit JBS.  
For example see https://bugs.openjdk.org/browse/JDK-8232933

yanglong1010 commented:
/approval 8312065 request I would like to fix this issue. 
This problem can be reproduced using pure Java code, or it can be reproduced by sending signal from an external profiling tool. If triggered, it may cause the thread to block abnormally. 
Low risk: The repair is clear，do not perform redundant loop retries for NET_Poll. Existing tests have proven that it does not affect the original functions, and a dedicated test for this issue has been added.
Testing: All GHA tests are passed. I ran all the jtreg tests (test/jdk/java/net/ServerSocket,test/jdk/java/net/Socket,test/jdk/java/net/SocketOption,test/jdk/java/net/ipv6) related to jdk.net.usePlainSocketImpl on my host, all passed. I've also added a test for this issue, the unfixed JDK fails, the fixed version passes.

yanglong1010 commented:
> You are still missing the fix request comment that states reason, risk and testing in the JBS bug. You can issue another approval request (Use: approval request " to do that, or edit JBS. For example see https://bugs.openjdk.org/browse/JDK-8232933

GoeLin, Thanks. Updated.

yanglong1010 commented:
/integrate

D-D-H commented:
/sponsor

D-D-H commented:
@yanglong1010 I'm not a committer of jdk-updates project :-(

phohensee commented:
/sponsor

yanglong1010 commented:
Thanks a lot, @apangin @vyommani @phohensee @GoeLin.
